cmd/cli/main.go

```diff
@@ -65,35 +73,54 @@ func checkEnv() {
 
 func validate() []error {
 	var errs []error
+
 	if accessToken == "" {
 		errs = append(errs, errors.New("error missing parameter. --token is required"))
 	}
 
+	if threescaleURL == "" {
+		errs = append(errs, errors.New("error missing parameter. --url is required"))
+	}
+
 	if svcID == "" {
 		errs = append(errs, errors.New("error missing parameter. --service is required"))
 	}
 
-	if threescaleURL == "" {
-		errs = append(errs, errors.New("error missing parameter. --url is required"))
-	}
 	return errs
 }
 
-func execute() error {
+func execute() []error {
+	var errs []error
 	var writeTo io.Writer
 
-	handler := templating.Handler{
-		AccessToken: accessToken,
-		SystemURL:   threescaleURL,
-		ServiceID:   svcID,
+	handler, errList := templating.NewHandler(accessToken, threescaleURL, svcID, fixup)
+	if len(errList) > 0 {
+		errs = append(errs, errList...)
 	}
 
 	instance := templating.GetDefaultInstance()
 	instance.AuthnMethod = templating.AuthenticationMethod(authType)
 
-	cg, err := templating.NewConfigGenerator(handler, instance, templating.Rule{})
-	if err != nil {
-		return fmt.Errorf("error - generating configuration - %s", err.Error())
+	if uid == "" {
+		// generate a UID from the URL
+		url, err := templating.ParseURL(handler.SystemURL)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("URL parsing for UID generation failed: %s", err.Error()))
+		} else {
+			uid, err = templating.UidGenerator(url, handler.ServiceID)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("UID generation failed: %s", err.Error()))
+			}
+		}
+	}
+
+	cg, errList := templating.NewConfigGenerator(*handler, instance, templating.Rule{}, templating.ConfigGeneratorOpts{UID: uid, Fixup: fixup})
```

Comment:
I think we always call this with an empty rule, don't we? We could remove it from the argument list and maybe we won't need to use that options struct.

Reply:
Right, yeah. Can't see that being a problem